tracing/tck/src/main/java/org/eclipse/microprofile/telemetry/tracing/tck/async/JaxRsServerAsyncTest.java

```diff
@@ -0,0 +1,144 @@
+                .addAsResource(config, "META-INF/microprofile-config.properties")
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml");
+    }
+
+    private static final String TEST_VALUE = "test.value";
+
+    @Inject
+    private Tracer tracer;
+
+    @Inject
+    private InMemorySpanExporter spanExporter;
+
+    @ArquillianResource
+    URL url;
+
+    @BeforeMethod
+    void setUp() {
+        // Only want to run on server
+        if (spanExporter != null) {
+            spanExporter.reset();
+        }
+    }
+
+    @Test(groups = "optional-jaxrs-tests")
+    public void testJaxRsServerAsyncCompletionStage() {
+        doAsyncTest(JaxRsServerAsyncTestEndpointClient::getCompletionStage);
+    }
+
+    @Test(groups = "optional-jaxrs-tests")
+    public void testJaxRsServerAsyncSuspend() {
+        doAsyncTest(JaxRsServerAsyncTestEndpointClient::getSuspend);
+    }
+
+    private void doAsyncTest(Function<JaxRsServerAsyncTestEndpointClient, String> requestFunction) {
+        Baggage baggage = Baggage.builder()
+                .put(JaxRsServerAsyncTestEndpoint.BAGGAGE_KEY, TEST_VALUE)
+                .build();
+
+        try (Scope s = baggage.makeCurrent()) {
+            // Make the request to the test endpoint
+            try {
+                JaxRsServerAsyncTestEndpointClient client = RestClientBuilder.newBuilder()
+                        .baseUri(url.toURI())
+                        .build(JaxRsServerAsyncTestEndpointClient.class);
+
+                String response = requestFunction.apply(client);
+                Assert.assertEquals("OK", response);
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        List<SpanData> spanData = spanExporter.getFinishedSpanItems(3);
+
+        // Assert correct parent-child links
+        // Shows that propagation occurred
+        SpanData subtaskSpan = spanExporter.getFirst(SpanKind.INTERNAL);
+        SpanData clientSpan = spanExporter.getFirst(SpanKind.CLIENT);
+        SpanData serverSpan = spanExporter.getFirst(SpanKind.SERVER);
+
+        assertEquals(serverSpan.getSpanId(), subtaskSpan.getParentSpanId());
+        assertEquals(clientSpan.getSpanId(), serverSpan.getParentSpanId());
+
+        // Assert that the expected headers were used
+        Assert.assertTrue(serverSpan.getAttributes().get(BAGGAGE_VALUE_ATTR).contains(TEST_VALUE));
+
+        // Assert baggage propagated on subtask span
+        Assert.assertTrue(subtaskSpan.getAttributes().get(BAGGAGE_VALUE_ATTR).contains(TEST_VALUE));
+
+        // Assert that the server span finished after the subtask span
+        // Even though the resource method returned quickly, the span should not end until the response is actually
+        // returned
+        Assert.assertTrue(serverSpan.getEndEpochNanos() > subtaskSpan.getEndEpochNanos());
```

Comment:
Strictly speaking, this should be a `>=` comparison.

It might also be helpful here to use `assertThat(serverSpan.getEndEpochNanos(), greaterThanOrEqualTo(subtaskSpan.getEndEpochNanos()))` so that any failure message includes the timestamps.